cmd/hosts-file-daemon/daemon.go

```diff
@@ -66,100 +67,177 @@ func Run(daemonConfig *DaemonConfig) {
 	go ManageIngressChanges(daemonConfig, updatesChannel, hostsFile)
 	go ManageServiceChanges(daemonConfig, updatesChannel, hostsFile)
 
+	go func() {
+		time.Sleep(time.Second * 60)
+		log.Println("Forcing update of hostsfile to ensure initial launch configurations persist")
+		hostfileContents := hostsFile.String()
+		err := WriteHostsFileAndRestartPihole(daemonConfig, hostfileContents)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}()
+
 	interrupt.WaitForAnySignal(syscall.SIGINT, syscall.SIGTERM)
 }
 
-func ManageIngressChanges(daemonConfig *DaemonConfig, updatesChannel chan *string, hosts *hostsfile.ConcurrentHostsFile) {
-	api := daemonConfig.KubernetesClientSet.ExtensionsV1beta1()
-	listOptions := metav1.ListOptions{}
-	watcher, err := api.Ingresses("").Watch(context.Background(), listOptions)
+func GetNginxIngress(obj interface{}) (*extensionsv1beta1.Ingress, *string, error) {
+	ingress, ok := obj.(*extensionsv1beta1.Ingress)
+	if !ok {
+		return nil, nil, errors.New(fmt.Sprintf("Failed to get ingress from provided object."))
+	}
 
-	if err != nil {
-		log.Fatal(err)
+	objectId := ingress.ObjectMeta.Namespace + "/" + ingress.ObjectMeta.Name
+
+	ingressClass, ok := ingress.Annotations["kubernetes.io/ingress.class"]
+	if !ok {
+		return nil, &objectId, errors.New(fmt.Sprintf("Skipping ingress (%s) because it doesn't have an ingress class\n", objectId))
+	}
+
+	if ingressClass != "nginx" {
+		return nil, &objectId, errors.New(fmt.Sprintf("Skipping ingress (%s) because it doesn't belong to NGINX Ingress Controller\n", objectId))
 	}
 
-	ch := watcher.ResultChan()
-	log.Println("Starting to monitor ingresses in all namespaces.")
+	return ingress, &objectId, nil
+}
 
-	for event := range ch {
-		ingress, ok := event.Object.(*extensionsv1beta1.Ingress)
-		if !ok {
-			log.Fatal("unexpected type")
+func UpdateHostsFromIngress(hosts *hostsfile.ConcurrentHostsFile, ingress *extensionsv1beta1.Ingress, objectId string, ingressIp string) bool {
+	hostnames := []string{}
+	for _, rule := range ingress.Spec.Rules {
+		if strings.HasSuffix(rule.Host, ingressIp) {
+			hostnames = append(hostnames, rule.Host+".")
+		} else {
+			hostnames = append(hostnames, rule.Host)
 		}
+	}
+	return hosts.SetHostnames(objectId, ingressIp, hostnames)
+}
+
+func ManageIngressChanges(daemonConfig *DaemonConfig, updatesChannel chan *string, hosts *hostsfile.ConcurrentHostsFile) {
+	// Resync every minute, just in case something somehow gets missed.
+	informerFactory := informers.NewSharedInformerFactory(daemonConfig.KubernetesClientSet, time.Minute)
```

Comment:
`informerFactory` could be passed in here and shared between the goroutine funcs.